fms/models/hf/mixtral/modeling_mixtral_hf.py

```diff
@@ -0,0 +1,125 @@
+from typing import Optional, Tuple
+
+import torch
+import torch.nn as nn
+from transformers import PretrainedConfig
+from transformers.modeling_outputs import BaseModelOutputWithPastAndCrossAttentions
+
+from fms.models.hf.lm_head_mixins import LMHeadModelLMHeadMixin
+from fms.models.hf.mixtral.configuration_mixtral_hf import HFAdaptedMixtralConfig
+from fms.models.hf.modeling_hf_adapter import HFDecoder, HFDecoderModelArchitecture
+from fms.models.mixtral import Mixtral, MixtralHeadless
+
+
+class HFAdaptedMixtralDecoder(HFDecoder):
+    """Adapter for the Mixtral decoder"""
+
+    def __init__(self, model: MixtralHeadless, config: PretrainedConfig):
+        super().__init__(model, config, attention_mask_dim=3)
+
+    def _adapt(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[Tuple[torch.Tensor]] = None,
+        use_cache: Optional[bool] = None,
+        attn_algorithm: Optional[
+            str
+        ] = None,  # this can be passed in from top most forward
+        *args,
+        **kwargs,
+    ) -> BaseModelOutputWithPastAndCrossAttentions:
+        output = self.model(
+            x=input_ids,
+            mask=attention_mask,
+            position_ids=position_ids,
+            past_key_value_states=past_key_values,
+            use_cache=use_cache,
+            attn_algorithm=attn_algorithm,
+        )
+
+        present_key_values = None
+        if isinstance(output, tuple):
+            output, present_key_values = output
+        return BaseModelOutputWithPastAndCrossAttentions(
+            last_hidden_state=output, past_key_values=present_key_values
+        )
+
+
+class HFAdaptedMixtralHeadless(HFDecoderModelArchitecture):
+    """This is the Adapter for the base Mixtral architecture"""
+
+    # attributes required by HF
+    config_class = HFAdaptedMixtralConfig
+    base_model_prefix = "hf_adapted_mixtral"
+
+    _tied_weights_keys = ["decoder.model.embedding.weight", "embedding.weight"]
+    _keys_to_ignore_on_save = ["embedding.weight"]
+
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        decoder: Optional[nn.Module] = None,
+        embedding: Optional[nn.Module] = None,
+        *args,
+        **kwargs,
+    ):
+        # in the case we have not yet received the encoder/decoder/embedding, initialize it here
+        if decoder is None or embedding is None:
+            params = config.to_dict()
+            model = MixtralHeadless(**params)
+            decoder = model if decoder is None else decoder
+            embedding = model.embedding if embedding is None else embedding
+
+        # these are now huggingface compatible
+        decoder = HFAdaptedMixtralDecoder(decoder, config)
+        super().__init__(decoder, embedding, config, *args, **kwargs)
+
+    def _prepare_inputs_for_generation(
```

Comment:
I'm wondering if we still need this, might be best to add it to the `HFDecoderModelArchitecture` if its generalizable

Reply:
I think it's mostly generalizable for FMS models, but if someone that's not FMS wants to use maybe they don't use "x" or "mask" as parameter names

Reply:
also, we do the Rope fix here because Mixtral also uses RoPE, but other models might not